### ISSUE TITLE
Surprising output from `recognize!` with `take_while!`

### ISSUE THREAD

**Issue by tari** [Geal/nom#225]
Not sure if there's a bug somewhere in here, but I've been trying to pin down the problem for several hours and made little progress.

Sample:
```
named!(ex, recognize!(preceded!(alpha, take_while!(is_digit))));
assert_eq!(ex(b"ab-"), IResult::Done(&b"-"[..], &b"ab"[..]));
assert_eq!(ex(b"a12"), IResult::Done(&b""[..], &b"a12"[..]));
```
The second assert fails with actual result `Done([], [97])`, having consumed only one byte of input but  leaving no unparsed data. I expect all of the input to be consumed and returned.

I've experimented with a number of variations on "two combinators inside `recognize`" but have only observed this kind of output when `take_while` is present.

---

For reference, the actual thing I'm trying to match is (substantially) one alphabetic character followed by zero or more alphanumeric characters. Perhaps `recognize` is the wrong tool for this job.

**Comment by tari** [comment]
This appears to be a result of how the `None` case from `position` is handled. An empty slice from the beginning of the input is taken rather than from the end, so the `offset` as taken by `recognize` is as if there were no input consumed.

I'm not sure if this behavior is by design, but examining `take_till` it also looks like it will behave unusually with `recognize`:
```
      match $input.iter().position(|c| $submac!(c, $($args)*)) {
        Some(n) => $crate::IResult::Done(&$input[n..], &$input[..n]),
        None    => $crate::IResult::Done(&b""[..], $input)
      }
```
In the `None` case an unrelated slice is returned for remaining input, which will have an arbitrary unknown offset into the input since the implementation of `offset` performs pointer arithmetic with the start of the slice.

**Comment by Geal** [comment]
Hi! This was fixed in 2927310. I planned to release nom 1.2.1 tomorrow, with this fix and other issues.

**Comment by Geal** [comment]
1.2.1 is now released
